### PR TITLE
[Merged by Bors] - feat: lemmas about `List.reverseRecOn`

### PR DESCRIPTION
**Pull request by eric-wieser**
This renames the arguments to the eliminators to be more ergonomic when using the `induction` tactic.

I also changed the definition in an attempt to make the proof easier.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/Basic.lean [hunk 1]
```diff
@@ -905,41 +905,106 @@ theorem modifyHead_modifyHead (l : List α) (f g : α → α) :
...
 prove statements about palindromes. The principle is given for a `Sort`-valued predicate, i.e., it
 can also be used to construct data. -/
-def bidirectionalRec {C : List α → Sort*} (H0 : C []) (H1 : ∀ a : α, C [a])
-    (Hn : ∀ (a : α) (l : List α) (b : α), C l → C (a :: (l ++ [b]))) : ∀ l, C l
-  | [] => H0
-  | [a] => H1 a
-  | a :: b :: l => by
+@[elab_as_elim]
+def bidirectionalRec {motive : List α → Sort*} (nil : motive []) (singleton : ∀ a : α, motive [a])
+    (cons_append : ∀ (a : α) (l : List α) (b : α), motive l → motive (a :: (l ++ [b]))) :
```
Comment (YaelDillies) [inline Mathlib/Data/List/Basic.lean:955]: If you want to shorten things a bit more without rendering them unreadable,
```suggestion
def bidirectionalRec {motive : List α → Sort*} (nil : motive []) (singleton : ∀ a, motive [a])
    (cons_append : ∀ a l b, motive l → motive (a :: (l ++ [b]))) :
```
(and below

Comment (eric-wieser) [reply Mathlib/Data/List/Basic.lean:955]: I don't think shortening them is necessary here; and at any rate, the verbosity is the same as it was before this patch.

**Review by YaelDillies — APPROVED** [review]
maintainer merge

**Comment by j-loreaux** [comment]
bors merge

**Pull request closed without merging**